controllers/product.go

```diff
@@ -0,0 +1,178 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"strconv"
+
+	"github.com/cloudinary/cloudinary-go/api/uploader"
+	"github.com/distributed-marketplace-system/db"
+	"github.com/distributed-marketplace-system/models"
+	"github.com/distributed-marketplace-system/util"
+	"github.com/gin-gonic/gin"
+)
+
+//ProductController ...
+type ProductController struct{}
+
+func UploadProductImage(img_path string) (url string, err error) {
+
+	if _, err := os.Stat(img_path); os.IsNotExist(err) {
+		return "", err
+	}
+
+	var ctx = context.Background()
+	uploadResult, err := util.CLD.Upload.Upload(ctx, img_path, uploader.UploadParams{Folder: "asu"})
+	if err != nil {
+		log.Fatalf("Failed to upload file, %v\n", err)
+		return "", err
+	}
+
+	log.Println(uploadResult.SecureURL)
+	return uploadResult.SecureURL, nil
+}
+
+func (ctrl ProductController) AddProduct(c *gin.Context) {
+
+	// Get the email of the user to compare with email in the header
+	var input models.AddProductInput
+	err := c.ShouldBind(&input)
+
+	var user models.User
+	if db.DB.Find(&user, "id=?", input.UserID).RecordNotFound() {
+		c.IndentedJSON(http.StatusBadRequest, gin.H{"err": "there is no user with id " + (string)(input.UserID)})
+		return
+	}
+
+	email := c.Request.Header.Get("email")
+	if email != user.Email {
+		c.JSON(http.StatusUnauthorized, gin.H{"error": "You are not authorized to be here"})
+		c.Abort()
+		return
+	}
+
+	if err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	// Extract image from the form
+	r := c.Request
+	r.ParseMultipartForm(10 << 20)
+	file, _, err := r.FormFile("image")
+	if err != nil {
+		fmt.Println("Error Retrieving the File")
+		fmt.Println(err)
+		return
+	}
+	defer file.Close()
+
+	// Store image in the server side
+	tempFile, err := ioutil.TempFile("res", "upload-*.png")
+	if err != nil {
+		fmt.Println(err)
+	}
+	defer tempFile.Close()
+
+	fileBytes, err := ioutil.ReadAll(file)
+	if err != nil {
+		fmt.Println(err)
+	}
+	// write this byte array to our temporary file
+	tempFile.Write(fileBytes)
+
+	// upload the image to cloudinary cloud
+	img_url, err := UploadProductImage(tempFile.Name())
+
+	if err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	// remove the image from the server side after uploading it
+	e := os.Remove(tempFile.Name())
+	if e != nil {
+		log.Fatal(e)
+	}
+
```

Comment:
This also to a separate function

Reply:
Also if we are going to be distributed we can try to run this async and update img_url later, but we don't need to for now